state-synchronizer/src/peer_manager.rs

```diff
@@ -122,52 +126,80 @@ impl PeerManager {
-                    error!(
-                        "[state sync] (pick_peer) failed to compute weighted index, {:?}",
-                        e
-                    );
-                }
-            }
-        }
+        // compute weighted index based on updated eligible peers
+        let mut eligible_peers = vec![];
+        let weights: Vec<_> = active_peers
+            .into_iter()
+            .map(|(peer, peer_info)| {
+                eligible_peers.push(peer.clone());
+                peer_info.score
+            })
+            .collect();
+        self.eligible_peers = eligible_peers;
+        self.weighted_index = WeightedIndex::new(&weights)
+            .or_else(|err| {
+                error!(
+                    "[state sync] (pick_peer) failed to compute weighted index, {:?}",
+                    err
+                );
+                Err(err)
+            })
+            .ok();
     }
 
     pub fn pick_peer(&self) -> Option<PeerNetworkId> {
-        let active_peers = self.get_active_upstream_peers();
-        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
-
         if let Some(weighted_index) = &self.weighted_index {
             let mut rng = thread_rng();
-            if let Some(peer) = active_peers.get(weighted_index.sample(&mut rng)) {
-                return Some(peer.0.clone());
+            if let Some(peer) = self.eligible_peers.get(weighted_index.sample(&mut rng)) {
+                return Some(peer.clone());
             }
         }
         None
     }
 
     fn get_active_upstream_peers(&self) -> Vec<(&PeerNetworkId, &PeerInfo)> {
-        self.peers
-            .iter()
-            .filter(|&(peer, peer_info)| peer_info.is_alive && self.is_upstream_peer(peer))
-            .collect()
+        if self.upstream_config.networks.len() > 1 {
+            // failover mode is enabled only if there are multiple upstream networks
+            // in failover mode, we select the network of the highest preference (defined by UpstreamConfig)
+            // with at least one live peer
+            // We failover to the next network (in order of preference defined by UpstreamConfig) if
+            // there are no live peers in each network
+
+            // group active upstream peers by network
+            let active_peers_by_network = self
+                .peers
+                .iter()
+                .filter(|(_peer, peer_info)| peer_info.is_alive)
+                .map(|(peer, peer_info)| (peer.network_id(), (peer, peer_info)))
+                .into_group_map();
+
+            // find the first network with any live peers
+            self.upstream_config
+                .networks
+                .iter()
+                .find_map(|network| active_peers_by_network.get(network))
+                .unwrap_or(&vec![])
+                .to_vec()
+        } else {
+            // no failover option
+            // all upstream peers belong to the same network
+            self.peers
```

Comment:
do you need this else branch? isn't it the same as above?

Reply:
for validators, its `NetworkId::Validator` is not explicitly specified in the upstream config, so the above branch would give no active peers

Reply:
hmm, when there's failover, `validator` would be specified in the upstream config as first element?

Reply:
I don't think there ever is a use case where a validator will ever failover? and a FN would never list the validator entwork as an upstream config

Reply:
can you give examples how the upstream config looks like as validator, FN, and FN with failover?

Reply:
validator: empty []
VFN w/o failover: [NetworkId::Private("vfn")] 
VFN w/ failover: [NetworkId::Private("vfn"), NetworkId::Public]
public FN: [NetworkId::Public]